### PR TITLE
fix issues #149 and #150 i.e. performance improvement of indexing and…

### LINKED ISSUE #149: improve indexing performance by using String and not Resource in CollectionFolder
so it can be used immediately as a string

### LINKED ISSUE #150: Support IndexService.indexUris where it receives uris which should not be indexed
Make the current implementation smarter, so we can pass in a list of URIs to be indexed and the component is smart enough to only index the URIs which match the criteria using the same `select-*` queries as the full reindex.

### PR DESCRIPTION
**Pull request by natan-cox-cognizone**
… smarter indexUris method to detect invalid uris

### CODE REVIEW

**Comment by natan-cox-cognizone** [comment]
fix #149

**Comment by natan-cox-cognizone** [comment]
fix #150

**Review by ssamarid — APPROVED** [review]

File: cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java [hunk 1]
```diff
@@ -140,6 +145,71 @@ static List<String> getUrisFromQuery(@Nonnull IndexingServiceContext context,
...
+      log.info("(index uri) ignoring {} of {} uris for indexing", uris.size() - indexableUris.size(), uris.size());
+    }
+
+    return indexableUris;
+  }
+
+  private static boolean shouldCheckForIndexableUris(@Nonnull CollectionFolder collectionFolder) {
+    return collectionFolder.getSelectQueries()
+                           .stream()
+                           .anyMatch(query -> query.contains("#{[uriFilter]}"));
```
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java:182]: scary

File: cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java [hunk 2]
```diff
@@ -140,6 +145,71 @@ static List<String> getUrisFromQuery(@Nonnull IndexingServiceContext context,
...
+                                       CollectionFolder collectionFolder,
+                                       List<String> uris) {
+    if (!shouldCheckForIndexableUris(collectionFolder))
+      return uris;
+
+
+    List<List<String>> uriSubLists = Lists.partition(uris, 20);
+    List<String> indexableUris =
+            uriSubLists.stream()
+                       .flatMap(uriSubList -> selectIndexableUris(context, collectionFolder, uriSubList).stream())
```
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java]: split in  (! import of Collection is needed here)
```suggestion
                       .map(uriSubList -> selectIndexableUris(context, collectionFolder, uriSubList))
                       .flapMap(Collection::stream)
```

File: cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java [hunk 3]
```diff
@@ -140,6 +145,71 @@ static List<String> getUrisFromQuery(@Nonnull IndexingServiceContext context,
...
+                           .stream()
+                           .anyMatch(query -> query.contains("#{[uriFilter]}"));
+  }
+
+  private static List<String> selectIndexableUris(@Nonnull IndexingServiceContext context,
+                                                  @Nonnull CollectionFolder collectionFolder,
+                                                  @Nonnull List<String> uris) {
+    return collectionFolder.getSelectQueries()
+                           .stream()
+                           .flatMap(query -> selectIndexableUris(context, query, uris).stream())
```
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java]: split also?

File: cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java [hunk 4]
```diff
@@ -44,22 +49,21 @@ static List<String> getCollectionUris(@Nonnull IndexingServiceContext context,
...
-                           .map(query -> paginatedQuery.select(rdfStore, query))
-                           .flatMap(queryResult -> paginatedQuery.convertSingleColumnUriToStringList(queryResult)
-                                                                   .stream())
-                           .collect(Collectors.toList());
+    return collectionFolder
+            .getSelectQueries()
+            .stream()
+            .map(query -> spelService.processTemplate(query, context.getQueryTemplateParameters()))
+            .map(query -> paginatedQuery.select(rdfStore, query))
+            .flatMap(queryResult -> paginatedQuery.convertSingleColumnUriToStringList(queryResult).stream())
```
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java:59]:
```suggestion
            .flatMap(paginatedQuery::convertSingleColumnUriToStringList)
            .flatMap(Collection::stream)
```

File: cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java (hunk 3 above)
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/InternalIndexingServiceUtils.java]:
```suggestion
                           .map(query -> selectIndexableUris(context, query, uris))
                           .flatMap(Collection::stream)
```

**Review by domivds — APPROVED** [review]

**Pull request merged**